fuel-poa-coordinator/src/service.rs

```diff
@@ -0,0 +1,326 @@
+            None
+        }
+    }
+}
+
+pub struct Task<S>
+where
+    S: InterpreterStorage + Send,
+    <S as InterpreterStorage>::DataError: Send,
+{
+    stop: mpsc::Receiver<()>,
+    block_gas_limit: Word,
+    db: S,
+    block_producer: Arc<dyn BlockProducer>,
+    txpool_sender: TxPoolSender,
+    txpool_broadcast: broadcast::Receiver<TxStatusBroadcast>,
+    import_block_events_tx: broadcast::Sender<ImportBlockBroadcast>,
+    /// Last block creation time. When starting up, this is initialized
+    /// to `Instant::now()`, which delays the first block on startup for
+    /// a bit, but doesn't cause any other issues.
+    last_block_created: Instant,
+    trigger: Trigger,
+    /// Deadline clock, used by the triggers
+    timer: DeadlineClock,
+}
+impl<S> Task<S>
+where
+    S: InterpreterStorage + Send,
+    <S as InterpreterStorage>::DataError: Send,
+{
+    // Request the block producer to make a new block, and return it when ready
+    async fn signal_produce_block(&mut self) -> anyhow::Result<FuelBlock> {
+        let current_height = self
+            .db
+            .block_height()
+            .map_err(|err| anyhow::format_err!("db error {err:?}"))?;
+        let height = BlockHeight::from(current_height + 1);
+
+        self.block_producer
+            .produce_block(height, self.block_gas_limit)
+            .await
+    }
+
+    async fn produce_block(&mut self) -> anyhow::Result<()> {
+        // Ask the block producer to create the block
+        let block = self.signal_produce_block().await?;
+
+        // Send the block back to the txpool
+        // TODO: this probably must be done differently with multi-node configuration
+        self.import_block_events_tx
+            .send(ImportBlockBroadcast::PendingFuelBlockImported {
+                block: Arc::new(block),
+            })
+            .expect("Failed to import the generated block");
+
+        // Update last block time
+        self.last_block_created = Instant::now();
+
+        // Set timer for the next block
+        match self.trigger {
+            Trigger::Never => {
+                unreachable!("This mode will never produce blocks");
+            }
+            Trigger::Instant => {}
+            Trigger::Interval { block_time } => {
+                // TODO: instead of sleeping for `block_time`, subtract the time we used for processing
+                self.timer.set_timeout(block_time, OnConflict::Min).await;
+            }
+            Trigger::Hybrid {
+                max_block_time,
+                min_block_time,
+                max_tx_idle_time,
+            } => {
+                self.timer
+                    .set_timeout(max_block_time, OnConflict::Min)
+                    .await;
+
+                // TODO: Maybe there could be a way to get the length without having
+                //       to perform full block production computation?
+                let block_tx_count =
```

Comment:
Since we are moving the selection of transactions to the txpool, we shouldn't need to use the block producer at all here.

Following from the P&A architecture proposal, the POA module could define it's own trait for getting exactly what it wants from the txpool. Since the POA coordinator only caree about whether total max gas consumable by all the txs in the pool is > the max gas per block, it could have a port like:

```
trait TransactionPool {
    fn get_total_comsumable_gas(&self) -> u64;
}
```

And then use it here to check: `self.txpool.get_total_comsumable_gas() > config.max_gas_per_block()`.

This way, the Poa doesn't risk producing extra or empty blocks for no reason.

Reply:
I now have local version that does exactly this. However, introducing additional channels makes the test timings even harder, and now some of the test cases are flaky, The flakyness is rather hard to work around, for the fix see this tokio issue https://github.com/tokio-rs/tokio/issues/4522  (and also tokio PR 4523, for an earlier attempt to fix this)